Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs

```diff
@@ -14,113 +14,394 @@
+                return false;
+            }
+            var successful = true;
+            foreach (var registration in registrationMap.Values)
+            {
+                var connection = registration.ClientConnection;
+                try
                 {
-                    Logger.Warning("Could not find the registration id alias for " + registrationId);
-                    return false;
-                }
+                    var serverRegistrationId = registration.ServerRegistrationId;
+                    var request = encodeDeregisterListenerRequest(serverRegistrationId);
 
-                var request = requestEncoder(realRegistrationId);
-                var task = _client.GetInvocationService().InvokeOnRandomTarget(request);
-                var actualResult = responseDecoder(ThreadUtil.GetResult(task));
-                if (Logger.IsFinestEnabled() && !actualResult)
+                    var future = ((ClientInvocationService) _client.GetInvocationService())
+                        .InvokeOnConnection(request, connection);
+                    ThreadUtil.GetResult(future);
+                    DistributedEventHandler removed;
+                    _eventHandlers.TryRemove(registration.CorrelationId, out removed);
+                    EventRegistration reg;
+                    registrationMap.TryRemove(connection, out reg);
+                }
+                catch (Exception e)
                 {
-                    Logger.Finest("Remove listener response returned false from server for registration id " +
-                                  registrationId);
+                    if (connection.Live)
+                    {
+                        successful = false;
+                        Logger.Warning(
+                            string.Format("Deregistration of listener with ID  {0} has failed to address {1}",
+                                userRegistrationId, connection.GetLocalSocketAddress()), e);
+                    }
                 }
-                return true;
+            }
+            if (successful)
+            {
+                _registrations.TryRemove(key, out registrationMap);
+            }
+            return successful;
+        }
+
+        private void RegisterListenerOnConnection(ListenerRegistrationKey registrationKey, ClientConnection connection)
+        {
+            //This method should only be called from registrationExecutor
+            Debug.Assert(!Thread.CurrentThread.Name.Contains("eventRegistration"));
+
+            ConcurrentDictionary<ClientConnection, EventRegistration> registrationMap;
+            if (_registrations.TryGetValue(registrationKey, out registrationMap) &&
+                registrationMap.ContainsKey(connection))
+            {
+                return;
+            }
+            var future = ((ClientInvocationService) _client.GetInvocationService())
+                .InvokeListenerOnConnection(registrationKey.RegistrationRequest, registrationKey.EventHandler,
```

Comment:
Java does invokeUrgent here. I did not see urgent invocation at CSharp, does it exist?

Reply:
It's backpressure related call. C# doesn't have that yet.